Implementation/5 implement product service

stishchenko opened the pull request:
Implementation of product service by streams and loops

rdovgan committed:
Added IsProductService for implementation

rdovgan committed:
Added test classes for ProductService

Tishchenko666 committed:
Merge branch 'master' into implementation/5_Implement_ProductService

Tishchenko666 committed:
ProductServiceImpl was created to implement IsProductService. All method except the first use Stream api.
Method defineProductNames() use foreach loop to check null product and throw exception or add its names to list.
Method defineProductsWithCreatedState() remove null objects and filter only create-state-products.
Method calculateProductCountByState() return map with zero-values for null-list or throw exception for list with null-objects or group and count products by state. [BUG] line 118.
Method filterProductsByProvidedObject() return empty list for null-product-list; throw exception for null-product; or filter by state or filter by name or filter by both and return list. [BUG] line 164.

Tishchenko666 committed:
[BUG] line 118 fixed. In case "map doesn't get all states from list" add missing states with zero-count.
[BUG] line 164 fixed. in case "product with null name and state" return whole product list.

Tishchenko666 committed:
In method calculateProductCountByState() creating map by null-list was refactored by foreach loop.

Tishchenko666 committed:
#5 implementation of ProductService

NckGrs commented on src/main/java/team/asd/service/ProductServiceImpl.java:
```diff
@@ -0,0 +1,99 @@
+		List<String> productNames = new ArrayList<>();
+		for (IsProduct p : productList) {
+			if (p == null) {
+				throw new WrongProductException("Product is null");
```
You can add a separate validation method to check the product list for empty values and use this method in multiple places before actions with the collection. Or look for some null safe method in the CollectionUtils class.

Tishchenko666 committed:
Null-list checking is replaced by CollectionUtils.isEmpty().
The list checking on null-elements is extracted in method checkListContainsNull().

NckGrs commented on src/main/java/team/asd/service/ProductServiceImpl.java:
```diff
@@ -0,0 +1,101 @@
+	}
+
+	private void checkListContainsNull(List<IsProduct> list) throws WrongProductException {
+		if (CollectionUtils.isNotEmpty(list)) {
```
you can combine two condition checks into one with &&.

Tishchenko666 committed:
If clauses for null-name or null-state was replaced by lambda in filter function of stream.

Tishchenko666 committed:
In method calculateProductCountByState() the creation of map with 0 values was replaced by null-safety in method countProductsWithState()

Tishchenko666 committed:
In method filterProductsByProvidedObject() filters were refactored to check one field per each.
